### PR TITLE
K8SPSMDB-335: Backup retention based on 'Keep' parameter

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-335](https://jira.percona.com/browse/K8SPSMDB-335) [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

This PR adds 'keep' field for backup task.
Based on that field operator will remove backup objects AND sent `deleteBackup` request to PBM to delete real backup data.

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| security-context | failed |
| arbiter | passed |
| operator-self-healing | passed |
| scheduled-backup | failed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| upgrade | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| self-healing | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | failed |
| security-context | failed |
| service-per-pod | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| rs-shard-migration | passed |
| smart-update | passed |
| version-service | passed |
| upgrade | failed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| monitoring-2-0 | failed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/apis/psmdb/v1/psmdb_types.go [hunk 1]
```diff
@@ -416,8 +416,9 @@ type MongodSpecOperationProfiling struct {
 }
 
 type BackupTaskSpec struct {
-	Name            string              `json:"name"`
-	Enabled         bool                `json:"enabled"`
+	Name            string              `json:"name,omitempty"`
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/apis/psmdb/v1/psmdb_types.go]: https://github.com/percona/percona-server-mongodb-operator/pull/539/commits/466be4cb00ef0f20c99ef239d34525fa09b308e2

File: pkg/controller/perconaservermongodb/backup.go [hunk 2]
```diff
@@ -64,3 +79,68 @@ func (r *ReconcilePerconaServerMongoDB) reconcileBackupTasks(cr *api.PerconaServ
...
+
+	h := &minHeap{}
+	heap.Init(h)
+	for _, bcp := range bcpList.Items {
+		if bcp.Status.State == api.BackupStateReady {
+			heap.Push(h, bcp)
+		}
+	}
+
+	if h.Len() <= keep {
```
Comment (qJkee) [inline pkg/controller/perconaservermongodb/backup.go:113]: we don't need this check, because we already checked at 101 line

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | failed |
| security-context | failed |
| operator-self-healing | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| rs-shard-migration | passed |
| smart-update | passed |
| upgrade | failed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| security-context | failed |
| operator-self-healing | passed |
| service-per-pod | passed |
| scheduled-backup | failed |
| rs-shard-migration | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | failed |

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| limits | passed |
| demand-backup | passed |
| scaling | passed |
| arbiter | passed |
| security-context | failed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | failed |
| scheduled-backup | failed |
| liveness | passed |
| smart-update | passed |
| demand-backup-sharded | passed |
| version-service | passed |
| upgrade | failed |
| users | passed |
| upgrade-sharded | failed |

**Pull request merged**